keps/sig-storage/3294-provision-volumes-from-cross-namespace-snapshots/README.md

```diff
@@ -0,0 +1,1065 @@
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+`VolumeSnapshot`s for PVCs in prod namespace are taken on a regular basis, PVCs are created from the `VolumeSnapshot`s in other test namespaces for testing.
+
+#### Story 2
+
+The same `VolumeSnapshot`s are expected to be consumed as golden images from multiple namespaces. Using PVs as [VM images for KubeVirt](https://kubevirt.io/2020/KubeVirt-VM-Image-Usage-Patterns.html) is one of the examples of this use case.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+#### Provisioning PVCs from cross-namespace PVCs
+
+The conclusion of the original discussion ([here](https://docs.google.com/document/d/17H1k4lqdtJwZSjNRaQue-FhMhyk14JA_MoURpoxha5Q/edit#bookmark=id.nj4e1ocn8b23) and [here](https://docs.google.com/document/d/17H1k4lqdtJwZSjNRaQue-FhMhyk14JA_MoURpoxha5Q/edit#bookmark=id.h1eqongxseo)) on transfer feature was that we should avoid implementing transfer of PVCs, because there will be more race conditions for PVCs than snapshots.
+However, we might have a room to reconsider if this cross-namespace-provision approach can solve the issue of race for PVCs, although transfer approach can't seem to resolve the issue easily.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Secret Handling
+
+Unlike transfer feature, this idea doesn't need to involve any transfers of Secert, therefore there will be no issue on Secret handling.
+From a populator, Secrets are only referenced through snapshots that exist in the same namespace (As commented [here](https://github.com/kubernetes/enhancements/pull/2849#issuecomment-962168202), depending on the driver implementation, there may be very little chance that some CSI drivers won't work well in a very rare situation. However, such drivers can avoid this issue separately, by turning off this feature, implementing their own populator, and so on).
+
+#### Security
+
+By using [`ReferencePolicy`](https://gateway-api.sigs.k8s.io/concepts/security-model/#2-referencepolicy), only allowed snapshots can be accessed beyond the namespace boundary (Please also see [original  discussion on security](https://github.com/kubernetes/enhancements/pull/2849#issuecomment-919107307)).
```

Comment:
To clarify, a ReferencePolicy would be required for a cross-namespace snapshot reference to be considered valid? Is this a point in time consideration? What happens if the ReferencePolicy that allowed this to be created is removed at a later point?

Reply:
> To clarify, a ReferencePolicy would be required for a cross-namespace snapshot reference to be considered valid? 

Yes.

> Is this a point in time consideration? 

Yes. It checks whether it is allowed when the controller is actually trying to use a cross-namespace snapshot.

> What happens if the ReferencePolicy that allowed this to be created is removed at a later point?

If it is allowed when controller is actually trying to use cross-namespace snapshot, the controller regard it as valid and start using it. If populating volume from snapshot succeeds without any errors, there will be no issue.

If it fails for some reason and the controller try to use the snapshot again and the ReferencePolicy is remove at that time, the controller will regard it as invalid and won't start using it until the valid ReferencePolicy is created.

I believe that this behavior is still consistent. Please point out if there are any corner cases to consider.

Reply:
That sounds fine to me, but will definitely need to be clearly documented in the release version; it seems likely that people might not think things will work in this way unless it's explained clearly.

Reply:
@youngnick Updated the KEP around `ReferenceGrant` in [this commit](https://github.com/kubernetes/enhancements/pull/3295/commits/74088d880846a55347d1f7e551a43bb7ead2fb92). PTAL